---
title: "ToleranceEvo_main"
author: "Álvaro Gómez Pérez"
date: "2025-07-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries
```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(data.table)
library(limpa)
library(helperfunctions)
library(preprocessCore)
library(proDA)
library(stringr)
library(glue)
library(factoextra)         # From here they are all for the heatmap/clustering section
library(dendextend)
library(staplr)
library(circlize)
library(ComplexHeatmap)
library(viridis)
```

Set up
```{r}
working_from <- "charite"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/"
} else
  if (working_from == "charite") {
    base_dir = "C:/MyStuff/"
  }
```


Load data
```{r}
# Precursor data wide format
raw_prec_data <- as.matrix(fread(paste(base_dir, "ToleranceEvo_Wenxi/Data/Preprocessing_steps/report_wide_filtered_norm_fully_imputed_drift_corrected_log2.csv", sep="")), rownames = 1)
hist(raw_prec_data)

# Metadata
metadata <- as.data.frame(fread(paste(base_dir, "ToleranceEvo_Wenxi/Data/metadata/metadata_processed.txt", sep="")))

# Correspondence between precursor IDs and protein IDs (UniProt) 
precursor_protein_correspondence <- as.data.frame(fread(paste(base_dir, "ToleranceEvo_Wenxi/Data/Preprocessing_steps/precursor_protein_correspondence.tsv", sep="")))

# UniProt dataframe
uniprot_db <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/databases/uniprotkb_proteome_UP000002311_2025_05_24.tsv", sep="")))
```




# 1. Differential expression analysis
Performed with limpa. Filtering and pre-processing at the precursor level was done in the cluster, here I start from the processed (filtered, normalized, drift corrected, MAR-only imputed) data at precursor level in wide format (precursors in rows, samples in columns), summarize to protein level and perform DE analysis, both with limpa.

## 1.1. Peptide to protein summarization
```{r}
# Prepare protein ID vector matching the precursor order in the rows of the raw data, setting the order based on that of the rows of the precursor raw data
precursor_protein_correspondence <- precursor_protein_correspondence %>%
  dplyr::filter(Precursor.Id %in% rownames(raw_prec_data))
precursor_protein_correspondence <- precursor_protein_correspondence[match(rownames(raw_prec_data), precursor_protein_correspondence$Precursor.Id),]


# Remove QC samples
raw_prec_data <- raw_prec_data[, !(grepl("QC", colnames(raw_prec_data)))]

# Run summarization
summarization_results <- protein_summarize(raw_data = raw_prec_data,
                                 protein_ids = precursor_protein_correspondence$Protein.Ids,
                                 method = "limpa")
protein_mat <- summarization_results$y.protein$E

# Remove proteins which are present in less than 65% of the samples per strain - NOT NECESSARY RIGHT NOW SINCE I'M USING FULLY IMPUTED DATASET
#protein_mat <- as.data.frame(protein_mat)
#protein_mat$Precursor.Id <- rownames(protein_mat)
#protein_mat <- pivot_longer(protein_mat,                                                  # Switch to longer format for filtering
#                                 cols = !Precursor.Id,
#                                 names_to = "Strain.Name",
#                                 values_to = "Protein.Intensity") %>%
#  dplyr::filter(!is.na(Protein.Intensity)) %>%                                            # Remove rows with NAs, otherwise this filtering doesn't work (it's based on the     #                                                                                          row being present or not, not on its contents)
#  remove_uncommon_precursors_per_strain(percentage_of_samples_per_precursor = 0.65) %>%   # Filter
#  pivot_wider(names_from = "Strain.Name",                                                 # Set back to wide format
#              values_from = "Protein.Intensity") %>%
#  as.data.frame() %>%
#  set_column_as_rownames("Precursor.Id") 
#sum(is.na(protein_mat))
#sum(is.na(protein_mat))/(ncol(protein_mat)*nrow(protein_mat))

# Save protein level dataset
saveRDS(summarization_results, paste(base_dir, "ToleranceEvo_Wenxi/Data/Preprocessing_steps/protein_summarized_fully_imputed.rds", sep = ""))
```



## 1.2. Differential expression analysis
### 1.2.1. Separately for each treatment
I run into the issue that lab strain at passage 0 in fluconazole was removed during pre-processing. I think that's possibly expected, since it's a lab strain and it's had no time to adapt, fluconazole will kill it, hence it's understandable that the proteome we detect in it is quite messy. Should check this with Wenxi though. 
```{r}
## Divide the dataset into 3 based on culture type (pellet, liquid, liquid+fluconazole) and perform DE analysis separate for each, using lab strain at passage 0 as reference
#treatments <- unique(metadata$Treatment)[!(unique(metadata$Treatment) == "QC")]
#da_list <- list()
#responsiveness_list <- list()
#
#for (treatment in treatments) {
#  # Grab the strains with this treatment
#  samples_to_select <- metadata$Sample.Name[metadata$Treatment == treatment]
#  temp_protein_mat <- protein_mat[, colnames(protein_mat) %in% samples_to_select]
#  reference_strain <- unique(metadata$Strain.Name[metadata$Treatment == treatment &
#                                             metadata$Passage == 0 &
#                                             metadata$Strain.Type == "Lab strain"])[1]
#  
#  # Run differential expression analysis
#  de_analysis_results <- de_analysis(raw_data = temp_protein_mat,
#                                     metadata = metadata,
#                                     method = "limma",
#                                     strain_name_column = "Strain.Name", 
#                                     sample_name_column = "Sample.Name", 
#                                     reference_strain = reference_strain,
#                                     N = 10,
#                                     lfc_threshold = 1,
#                                     alpha = 0.05)
#  da <- de_analysis_results$DE_df
#  da_list[[treatment]] <- da
#  
#  # Get responsiveness df
#  resp <- create_responsiveness_df(da,
#                                   lfc_threshold = 1,
#                                   alpha = 0.05)
#  responsiveness_list[[treatment]] <- resp
#}
#
#
#
```

### 1.2.2. All strains together, using C0L as the reference
Decided to use this one because it's the lab strain, at passage 0, and in liquid because it's supposed to have had a bit more time to grow than the solid one
```{r}
de_analysis_results_vs_C0L <- de_analysis(raw_data = summarization_results$y.protein,
                                   metadata = metadata,
                                   method = "limpa",
                                   strain_name_column = "Strain.Name", 
                                   sample_name_column = "Sample.Name", 
                                   reference_strain = "C0L",
                                   N = 10,
                                   lfc_threshold = 1,
                                   alpha = 0.05)
da_vs_C0L <- de_analysis_results_vs_C0L$DE_df

#resp <- create_responsiveness_df(da,                       # Didn't push this function from home :(
#                                 lfc_threshold = 1,
#                                 alpha = 0.05)

# This code is temporary until I get that function from home PC, then I go back to the lines above
alpha <- 0.05
lfc_threshold <- 1
resp_vs_C0L <- da_vs_C0L %>%
  group_by(Strain.Name) %>%
  summarise(nDEP = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha, na.rm = T),
            Up_regulated_adjusted = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha & diffexpressed_adjusted == "Up_regulated", na.rm = T),
            Down_regulated_adjusted = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha & diffexpressed_adjusted == "Down_regulated", na.rm = T),
            Up_regulated_non_adjusted = sum(abs(logFC) >= lfc_threshold & P.Value <= alpha & diffexpressed_non_adjusted == "Up_regulated", na.rm = T),
            Down_regulated_non_adjusted = sum(abs(logFC) >= lfc_threshold & P.Value <= alpha & diffexpressed_non_adjusted == "Down_regulated", na.rm = T),
            Replicate_num = mean(Replicate_num))
# Everything below this should work fine as well when I go back to that function

# Add other informative columns to resp df
temp <- metadata %>%
  dplyr::select(Strain.Name, Strain.Type, Treatment, Passage) %>%
  dplyr::distinct(Strain.Name, .keep_all = T)
resp_vs_C0L <- left_join(resp_vs_C0L, temp, by = "Strain.Name")
```

### 1.2.3. All strains together, using C0P as the reference
When using C0L as reference, I saw DEPs only in pellet and liquid+fluconazole, pointing towards the fact that it is the treatment that drives the proteome, not the strain nor the passage
```{r}
de_analysis_results_vs_C0P <- de_analysis(raw_data = summarization_results$y.protein,
                                   metadata = metadata,
                                   method = "limpa",
                                   strain_name_column = "Strain.Name", 
                                   sample_name_column = "Sample.Name", 
                                   reference_strain = "C0P",
                                   N = 10,
                                   lfc_threshold = 1,
                                   alpha = 0.05)
da_vs_C0P <- de_analysis_results_vs_C0P$DE_df

#resp <- create_responsiveness_df(da,                       # Didn't push this function from home :(
#                                 lfc_threshold = 1,
#                                 alpha = 0.05)

# This code is temporary until I get that function from home PC, then I go back to the lines above
alpha <- 0.05
lfc_threshold <- 1
resp_vs_C0P <- da_vs_C0P %>%
  group_by(Strain.Name) %>%
  summarise(nDEP = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha, na.rm = T),
            Up_regulated_adjusted = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha & diffexpressed_adjusted == "Up_regulated", na.rm = T),
            Down_regulated_adjusted = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha & diffexpressed_adjusted == "Down_regulated", na.rm = T),
            Up_regulated_non_adjusted = sum(abs(logFC) >= lfc_threshold & P.Value <= alpha & diffexpressed_non_adjusted == "Up_regulated", na.rm = T),
            Down_regulated_non_adjusted = sum(abs(logFC) >= lfc_threshold & P.Value <= alpha & diffexpressed_non_adjusted == "Down_regulated", na.rm = T),
            Replicate_num = mean(Replicate_num))
# Everything below this should work fine as well when I go back to that function

# Add other informative columns to resp df
temp <- metadata %>%
  dplyr::select(Strain.Name, Strain.Type, Treatment, Passage) %>%
  dplyr::distinct(Strain.Name, .keep_all = T)
resp_vs_C0P <- left_join(resp_vs_C0P, temp, by = "Strain.Name")
```



## 1.3. Boxplots for DE
### 1.3.1. DE vs. C0L
```{r}
# Per strain
ggplot(data = resp_vs_C0L, aes(x = Strain.Type, y = nDEP, group = Strain.Type, fill = Strain.Type)) +
  geom_boxplot() +
  geom_jitter(color="black", size=2, alpha=0.9) +
  theme_light() +
  xlab("Strain type") +
  ylab("nDEP") +
  labs(title = "Number of differentially expressed proteins vs. C0L",
       subtitle = "By strain type")

# Per passage
ggplot(data = resp_vs_C0L, aes(x = Passage, y = nDEP, group = Passage, fill = Passage)) +
  geom_boxplot() +
  geom_jitter(color="black", size=2, alpha=0.9) +
  theme_light() +
  xlab("Passage") +
  ylab("nDEP") +
  labs(title = "Number of differentially expressed proteins vs. C0L",
       subtitle = "By passage")

# Per treatment
ggplot(data = resp_vs_C0L, aes(x = Treatment, y = nDEP, group = Treatment, fill = Treatment)) +
  geom_boxplot() +
  geom_jitter(color="black", size=2, alpha=0.9) +
  theme_light() +
  xlab("Treatment") +
  ylab("nDEP") +
  labs(title = "Number of differentially expressed proteins vs. C0L",
       subtitle = "By treatment")
```

### 1.3.2. DE vs. C0P
```{r}
# Per strain
ggplot(data = resp_vs_C0P, aes(x = Strain.Type, y = nDEP, group = Strain.Type, fill = Strain.Type)) +
  geom_boxplot() +
  geom_jitter(color="black", size=2, alpha=0.9) +
  theme_light() +
  xlab("Strain type") +
  ylab("nDEP") +
  labs(title = "Number of differentially expressed proteins vs. C0P",
       subtitle = "By strain type")

# Per passage
ggplot(data = resp_vs_C0P, aes(x = Passage, y = nDEP, group = Passage, fill = Passage)) +
  geom_boxplot() +
  geom_jitter(color="black", size=2, alpha=0.9) +
  theme_light() +
  xlab("Passage") +
  ylab("nDEP") +
  labs(title = "Number of differentially expressed proteins vs. C0P",
       subtitle = "By passage")

# Per treatment
ggplot(data = resp_vs_C0P, aes(x = Treatment, y = nDEP, group = Treatment, fill = Treatment)) +
  geom_boxplot() +
  geom_jitter(color="black", size=2, alpha=0.9) +
  theme_light() +
  xlab("Treatment") +
  ylab("nDEP") +
  labs(title = "Number of differentially expressed proteins vs. C0P",
       subtitle = "By treatment")
```




## 1.4. Clustering analysis
Perform hierarchical clustering
```{r}
# Elbow method - try to figure out what number of clusters is appropriate 
clust_df <- protein_mat
clust_df <- as.data.frame(t(clust_df))         
clust_df <- scale(clust_df)

fviz_nbclust(clust_df, kmeans, method = "wss", k.max = 15)


# Create a function that can be used to perform clustering with different k values 
perform_clustering <- function(tmp, dist_method, clust_method, num_of_clusters, want_dendrogram = T, want_clust_plot = T) {
  # Get distance matrix based on Euclidean distance
  dist.euc <- dist(tmp, method = dist_method)
  
  # Perform clustering
  tree.complete <- hclust(dist.euc, method = clust_method)
  
  # Plot dendrogram
  if (want_dendrogram) {
    plot(tree.complete, cex = 0.3, 
         main = glue("Dendrogram for clustering with {clust_method} link"),
         xlab = glue("{dist_method} distance"))
    rect.hclust(tree.complete , k = num_of_clusters)
  }
  
  if (want_dendrogram) {
    dend_obj <- as.dendrogram(tree.complete)
    col_dend <- color_branches(dend_obj, k = num_of_clusters)
    labels_cex(col_dend)
    plot(col_dend)
  }
  
  # Visualize clusters
  my_clusters <- cutree(tree.complete, k = num_of_clusters)
  if (want_clust_plot) {print(fviz_cluster(list(data = tmp, cluster = my_clusters), 
                                           main = glue("Clustering by {clust_method} link"),
                                           ggtheme = theme_light()))}
  
  return(my_clusters)
}


# Use this function to do clustering: k = 2
my_clusters <- perform_clustering(clust_df, "euclidean", "complete", 6)
```

Clustering to save dendrogram to PDF
```{r}
# Set up parameters
distance_method <- "euclidean"
clustering_method <- "complete"

# Perform clustering and create dendrogram - CLUST_DF IS SCALED IN THE PREVIOUS CHUNK! IS THAT OKAY???
dist_mat <- dist(clust_df, method = distance_method)
tree <- hclust(dist_mat, method = clustering_method)
plot(tree, cex = 0.3)
hcd <- as.dendrogram(tree)

# Open a PDF for plotting; units are inches by default
pdf(paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/dendrograms/dist_method_",  distance_method, "_clust_method_", clustering_method, ".pdf", sep=""), width=40, height=15)

# Plot the tree into the PDF
plot(tree, cex = 0.3)

# Close the PDF file's associated graphics device (necessary to finalize the output)
dev.off()

# Rotate PDF
rotate_pdf(
  page_rotation = 90,
  input_filepath = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/dendrograms/dist_method_",  distance_method, "_clust_method_", clustering_method, ".pdf", sep=""),
  output_filepath = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/dendrograms/dist_method_",  distance_method, "_clust_method_", clustering_method, ".pdf", sep=""),
  overwrite = TRUE
)
```

Filter out proteins with low variability, towards the heatmaps
```{r}
protein_variability <- apply(protein_mat, 1, function(x) var(x, na.rm = T))
hist(protein_variability)
most_variable_proteins <- rownames(protein_mat)[protein_variability > 0.4]
trna_ko_high_var <- protein_mat[rownames(protein_mat) %in% most_variable_proteins,]
```


Heatmaps
- Heatmap with Z-score based on median per protein
So this is basically focusing on the variability of each protein across all the samples - removing the basal "effect" of that protein?
```{r}
create_heatmap(protein_mat = protein_mat,
               annotation = T,
               metadata = metadata,
               heatmap_type = "z-score",
               z_score_type = "standard",
               color_palette = "viridis",
               metadata_column_for_annotation = "Strain.Type",
               metadata_column_for_sample_name = "Sample.Name",
               location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/heatmaps/", sep=""),
               save_as = "pdf")

create_heatmap(protein_mat = protein_mat,
               annotation = T,
               metadata = metadata,
               heatmap_type = "z-score",
               z_score_type = "standard",
               color_palette = "viridis",
               metadata_column_for_annotation = "Treatment",
               metadata_column_for_sample_name = "Sample.Name",
               location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/heatmaps/", sep=""),
               save_as = "pdf")

create_heatmap(protein_mat = protein_mat,
               annotation = T,
               metadata = metadata,
               heatmap_type = "z-score",
               z_score_type = "standard",
               color_palette = "viridis",
               metadata_column_for_annotation = "Passage",
               metadata_column_for_sample_name = "Sample.Name",
               location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/heatmaps/", sep=""),
               save_as = "pdf")
```

- Heatmap with per-gene log2 FC across strains (compared to median of gene across all strains)
Here I'm normalizing the values for each protein by dividing them by the median abundance of that protein
```{r}
create_heatmap(protein_mat = protein_mat,
               annotation = T,
               metadata = metadata,
               heatmap_type = "log2FC",
               z_score_type = "standard",
               color_palette = "viridis",
               metadata_column_for_annotation = "Strain.Type",
               metadata_column_for_sample_name = "Sample.Name",
               location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/heatmaps/", sep=""),
               save_as = "pdf")

create_heatmap(protein_mat = protein_mat,
               annotation = T,
               metadata = metadata,
               heatmap_type = "log2FC",
               z_score_type = "standard",
               color_palette = "viridis",
               metadata_column_for_annotation = "Treatment",
               metadata_column_for_sample_name = "Sample.Name",
               location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/heatmaps/", sep=""),
               save_as = "pdf")

create_heatmap(protein_mat = protein_mat,
               annotation = T,
               metadata = metadata,
               heatmap_type = "log2FC",
               z_score_type = "standard",
               color_palette = "viridis",
               metadata_column_for_annotation = "Passage",
               metadata_column_for_sample_name = "Sample.Name",
               location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/heatmaps/", sep=""),
               save_as = "pdf")
```



## 1.5. PCA
I make 3 different ones, in each of them coloring by a different covariate
```{r}
do_pca(protein_mat,
       metadata,
       metadata_column_for_annotation = "Strain.Type",
       metadata_column_for_sample_name = "Sample.Name",
       location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/PCA/", sep=""),
       save_as = "pdf")

do_pca(protein_mat,
       metadata,
       metadata_column_for_annotation = "Passage",
       metadata_column_for_sample_name = "Sample.Name",
       location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/PCA/", sep=""),
       save_as = "pdf")

do_pca(protein_mat,
       metadata,
       metadata_column_for_annotation = "Treatment",
       metadata_column_for_sample_name = "Sample.Name",
       location_to_save = paste(base_dir, "ToleranceEvo_Wenxi/Output/Plots/PCA/", sep=""),
       save_as = "pdf")
```
















Debug
```{r}
raw_data = protein_mat$y.protein
metadata = metadata
method = "limma"
strain_name_column = "Strain.Name"
sample_name_column = "Sample.Name"
reference_strain = "C0L"
lfc_threshold = 1
alpha = 0.05
N = 10
```







































